### PR TITLE
Use only type hierarchy as an input to MergeInstrumentationAnalysisTransform

### PR DESCRIPTION
**Pull request by asodja**
Before if any local build-logic project was recompiled we re-run "merge analysis" step for whole classpath. So for all jars.
With this change "merge analysis" step is run only if type hiearachy changes.

This reduces newly generated files after:
```
git clean -ffdx && ./gradlew help --no-configuration-cache --no-build-cache
```

for `gradle/gradle` project in `caches/<version>/transforms/` from ~20000 to ~300. Gradle 8.7 produces 18 new files in transforms cache, but might also produce files in other directories.

~~Another improvement could be, that we would use `@Classpath` for "analyze artifact" step: that brings newly created files to ~200. But might have some performance penalty, so I am a bit reluctant to add. I would vote to try that option if improvement above really won't be sufficient.~~

Actually, using classes variant reduces the newly created files to ~200.

Mostly fixes https://github.com/gradle/gradle-private/issues/4162

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/registration/InstrumentationTransformRegisterer.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ private void registerInstrumentationAndUpgradesPipeline(long contextId, Dependen
                 spec.parameters(params -> {
                     params.getBuildService().set(service);
                     params.getContextId().set(contextId);
-                    params.getAnalysisResult().setFrom(service.map(it -> it.getAnalysisResult(contextId)));
+                    params.getTypeHierarchyAnalysis().setFrom(service.map(it -> it.getTypeHierarchyAnalysis(contextId)));
```
Comment (asodja) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/registration/InstrumentationTransformRegisterer.java:88]: 💭 Here we set type hierarchy analysis as an input.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle clean

**Review by lptr — APPROVED** [review]
LGTM.

**Review by lptr — APPROVED** [review]

**Comment by asodja** [comment]
@bot-gradle test this

**Pull request merged**